### ISSUE TITLE
[perf] GPU-resident inference: eliminate per-kernel CPU↔GPU round-trips

### ISSUE THREAD

**Issue by ziliangpeng** [ziliangpeng/triton-llm#32]
## Problem

Current architecture does ~15 CPU↔GPU round-trips per layer per decode step (420+ per step for SmolLM2-135M 30 layers):

```
for each kernel:
  gpu.to_device(input)    → CPU→GPU copy
  kernel_launch           → ~10-50µs Triton launch overhead
  gpu.synchronize()       → stall, wait for GPU
  gpu.to_host(output)     → GPU→CPU copy
```

Each round-trip adds ~30-70µs overhead. For 420 calls × 100 decode tokens = ~1.2-3s just in overhead per generation.

## Root cause

Three independent but compounding issues:

### 1. Weights copied CPU→GPU on every call
Every gemm() call is (np.array, np.array) → np.array. It to_device()s the weight matrix every time, even though weights never change between calls. They should be allocated on GPU once at init time.

### 2. Inter-kernel sync is unnecessary
Each kernel does synchronize() + to_host(). The calling code then immediately calls the next kernel with the result. The sync + copy-back + re-copy adds latency but zero value — the GPU could just pass device pointers directly.

### 3. KV cache lives on CPU
Cache is allocated as np.zeros() on CPU memory. Every decode step does a numpy copy (not a view), then to_device() copies it again to GPU.

### Phase 1: GPU-resident weights
Store weight matrices on GPU once at init time. Create device-pointer variants of kernel launchers that skip to_device/to_host.

### Phase 2: GPU KV cache
Allocate KV cache directly on GPU memory. Slice writes use device-pointer arithmetic.

### Phase 3: Fused decode loop
Single decode step runs entirely on GPU — only copy final logits back for sampling.

## Expected impact

| Bottleneck | Before | After |
|------------|--------|-------|
| Per-kernel CPU copy | 420× per step | 1× at end |
| KV cache copy | ~256KB/step | 0 |
| Weight re-upload | ~99% of calls | 0 |
| Est. TPOT (FP32) | ~250ms | ~5-15ms |
| Est. TPOT (FP16) | — | ~1-2ms |

No new kernels needed — just device-pointer variants of existing launchers.
